File: fancywebview/src/main/java/co/fitcom/fancywebview/AdvancedWebView.java
package co.fitcom.fancywebview;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.browser.customtabs.CustomTabsCallback;
import androidx.browser.customtabs.CustomTabsClient;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.browser.customtabs.CustomTabsServiceConnection;
import androidx.browser.customtabs.CustomTabsSession;

/**
 * Created by triniwiz on 1/18/18.
 */

public class AdvancedWebView {
    private static CustomTabsServiceConnectionCallBack customTabsServiceConnection;
    private static CustomTabsCallbackListener customTabsCallbackListener;
    private static final String PACKAGE_NAME = "com.android.chrome";
    public static final int REQUEST_CODE = 1868;

    private AdvancedWebViewListener webViewListener;
    private CustomTabsClient customTabsClient;
    private boolean warmUpView = false;
    private CustomTabsSession customTabsSession;
    private CustomTabsIntent customTabsIntent;
    private CustomTabsIntent.Builder builder;
    private Context mContext;

    public AdvancedWebView(Context context, @Nullable AdvancedWebViewListener listener) {
        mContext = context;
        setWebViewListener(listener);
        setUp();
    }

    void setUp() {
        if (customTabsServiceConnection.getCustomTabsClient() != null) {
            customTabsClient = customTabsServiceConnection.getCustomTabsClient();
            customTabsSession = customTabsClient.newSession(customTabsCallbackListener);
            builder = new CustomTabsIntent.Builder(customTabsSession);
        }
    }

    public static void init(Context context, boolean warmUp) {
        if (customTabsServiceConnection == null) {
            customTabsServiceConnection = new CustomTabsServiceConnectionCallBack(null, warmUp);
        }

        if (customTabsCallbackListener == null) {
            customTabsCallbackListener = new CustomTabsCallbackListener(null);
        }

        CustomTabsClient.bindCustomTabsService(context, PACKAGE_NAME, customTabsServiceConnection);
    }

    public void setWebViewListener(AdvancedWebViewListener listener) {
        webViewListener = listener;
        customTabsServiceConnection.setWebViewListener(listener);
        customTabsCallbackListener.setWebViewListener(listener);
    }

    public void setCustomTabsClient(CustomTabsClient client) {
        customTabsClient = client;
    }

    public CustomTabsClient getCustomTabsClient() {
        return customTabsClient;
    }

    public void loadUrl(String string) {
        PackageManager pm = mContext.getPackageManager();
        try {
            ApplicationInfo info = pm.getApplicationInfo(PACKAGE_NAME, 0);
            if (info.enabled) {
                setUp();
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                if (builder != null) {
                    customTabsIntent = builder.build();
                    customTabsIntent.intent.setPackage(PACKAGE_NAME);
                }
            }
            if (customTabsIntent != null) {
                customTabsIntent.launchUrl(mContext, Uri.parse(string));
            } else {
                WebViewFallback fallback = new WebViewFallback();
                fallback.openUri((Activity) mContext, Uri.parse(string));
            }
        } catch (PackageManager.NameNotFoundException e) {
            WebViewFallback fallback = new WebViewFallback();
            fallback.openUri((Activity) mContext, Uri.parse(string));
        }
    }

    public CustomTabsSession getSession() {
        return customTabsSession;
    }

    public CustomTabsIntent.Builder getBuilder() {
        return builder;
    }

}


class CustomTabsServiceConnectionCallBack extends CustomTabsServiceConnection {
    private AdvancedWebViewListener webViewListener;
    private boolean warmUpView = false;
    private CustomTabsClient mClient;

    public CustomTabsServiceConnectionCallBack(@Nullable AdvancedWebViewListener listener, boolean warmUp) {
        webViewListener = listener;
        warmUpView = warmUp;
    }

    @Override
    public void onCustomTabsServiceConnected(ComponentName name, CustomTabsClient client) {
        if (warmUpView) {
            client.warmup(0L);
        }
        if (webViewListener != null) {
            webViewListener.onCustomTabsServiceConnected(name, client);
        }
        mClient = client;
    }

    @Override
    public void onServiceDisconnected(ComponentName name) {
        if (webViewListener != null) {
            webViewListener.onServiceDisconnected(name);
        }
    }

    public void setWebViewListener(AdvancedWebViewListener listener) {
        webViewListener = listener;
    }

    public CustomTabsClient getCustomTabsClient() {
        return mClient;
    }
}


class CustomTabsCallbackListener extends CustomTabsCallback {
    private AdvancedWebViewListener webViewListener;

    public CustomTabsCallbackListener(@Nullable AdvancedWebViewListener listener) {
        webViewListener = listener;
    }

    @Override
    public void onNavigationEvent(int navigationEvent, Bundle extras) {
        if (webViewListener != null) {
            webViewListener.onNavigationEvent(navigationEvent, extras);
        }
    }

    public void setWebViewListener(AdvancedWebViewListener listener) {
        webViewListener = listener;
    }

}

File: fancywebview/src/main/java/co/fitcom/fancywebview/AdvancedWebView.kt
package co.fitcom.fancywebview

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import androidx.browser.customtabs.*

/**
 * Created by triniwiz on 1/18/18.
 */


object AdvancedWebViewStatics {
    var customTabsServiceConnection: CustomTabsServiceConnectionCallBack? = null
    var customTabsCallbackListener: CustomTabsCallbackListener? = null
    const val PACKAGE_NAME = "com.android.chrome"
    const val REQUEST_CODE = 1868

    fun init(context: Context, warmUp: Boolean) {
        if (customTabsServiceConnection == null) {
            customTabsServiceConnection = CustomTabsServiceConnectionCallBack(null, warmUp)
        }

        if (customTabsCallbackListener == null) {
            customTabsCallbackListener = CustomTabsCallbackListener(null)
        }

        CustomTabsClient.bindCustomTabsService(context, PACKAGE_NAME, customTabsServiceConnection)
    }
}

class AdvancedWebView(private val mContext: Context, listener: AdvancedWebViewListener?) {

    private var webViewListener: AdvancedWebViewListener? = null
    private var customTabsClient: CustomTabsClient? = null
    private var session: CustomTabsSession? = null
    private var customTabsIntent: CustomTabsIntent? = null
    private var builder: CustomTabsIntent.Builder? = null

    init {
        setWebViewListener(listener)
        setUp()
    }


    private fun setUp() {
        if (AdvancedWebViewStatics.customTabsServiceConnection!!.customTabsClient != null) {
            customTabsClient = AdvancedWebViewStatics.customTabsServiceConnection!!.customTabsClient
            session = customTabsClient!!.newSession(AdvancedWebViewStatics.customTabsCallbackListener)
            builder = CustomTabsIntent.Builder(session)
        }
    }

    private fun setWebViewListener(listener: AdvancedWebViewListener?) {
        webViewListener = listener
        AdvancedWebViewStatics.customTabsServiceConnection!!.setWebViewListener(listener)
        AdvancedWebViewStatics.customTabsCallbackListener!!.setWebViewListener(listener)
    }

    fun loadUrl(string: String) {
        val pm = mContext.packageManager
        try {
            val info = pm.getApplicationInfo(AdvancedWebViewStatics.PACKAGE_NAME, 0)

            if (info.enabled) {
                if (builder != null) {
                    customTabsIntent = builder!!.build()
                    customTabsIntent!!.intent.setPackage(AdvancedWebViewStatics.PACKAGE_NAME)
                }
            }
            if (customTabsIntent != null) {
                customTabsIntent!!.launchUrl(mContext, Uri.parse(string))
            } else {
                val fallback = WebViewFallback()
                fallback.openUri(mContext as Activity, Uri.parse(string))
            }
        } catch (e: PackageManager.NameNotFoundException) {
            val fallback = WebViewFallback()
            fallback.openUri(mContext as Activity, Uri.parse(string))
        }

    }

    fun getBuilder(): CustomTabsIntent.Builder? {
        return builder
    }

    fun setBuilder(builder: CustomTabsIntent.Builder?) {
        if (builder != null) this.builder = builder
    }

}


class CustomTabsServiceConnectionCallBack(private var webViewListener: AdvancedWebViewListener?, warmUp: Boolean) : CustomTabsServiceConnection() {
    private var warmUpView = false
    var customTabsClient: CustomTabsClient? = null
        private set

    init {
        warmUpView = warmUp
    }

    override fun onCustomTabsServiceConnected(name: ComponentName, client: CustomTabsClient) {
        if (warmUpView) {
            client.warmup(0L)
        }
        if (webViewListener != null) {
            webViewListener!!.onCustomTabsServiceConnected(name, client)
        }
        customTabsClient = client
    }

    override fun onServiceDisconnected(name: ComponentName) {
        if (webViewListener != null) {
            webViewListener!!.onServiceDisconnected(name)
        }
    }

    fun setWebViewListener(listener: AdvancedWebViewListener?) {
        webViewListener = listener
    }
}


class CustomTabsCallbackListener(private var webViewListener: AdvancedWebViewListener?) : CustomTabsCallback() {

    override fun onNavigationEvent(navigationEvent: Int, extras: Bundle?) {
        if (webViewListener != null) {
            webViewListener!!.onNavigationEvent(navigationEvent, extras)
        }
    }

    fun setWebViewListener(listener: AdvancedWebViewListener?) {
        webViewListener = listener
    }

}
